### PR TITLE
up the priority of numpy array comparisons in self.assertEqual

### LINKED ISSUE #58988: `exact_dtype=True` is ignored when comparing `torch.Tensor`'s and `np.ndarray`'s
## To Reproduce

```python
import torch
import numpy as np

from torch.testing._internal.common_utils import TestCase


class TestFoo(TestCase):
    def test_bar(self):
        x = torch.ones((1,), dtype=torch.float16)
        y = np.ones((1,), dtype=np.float32)

        self.assertEqual(x, y, exact_dtype=True)
```

## Expected behavior

Test should fail, but  passes.

## Environment

```
PyTorch version: 1.9.0a0+gitda7bfa1
Is debug build: True
CUDA used to build PyTorch: Could not collect
ROCM used to build PyTorch: N/A

OS: Arch Linux (x86_64)
GCC version: (crosstool-NG 1.24.0.133_b0863d8_dirty) 9.3.0
Clang version: Could not collect
CMake version: version 3.20.2
Libc version: glibc-2.9

Python version: 3.6 (64-bit runtime)
Python platform: Linux-5.12.5-arch1-1-x86_64-with-arch
Is CUDA available: False
CUDA runtime version: 11.1.105
GPU models and configuration: GPU 0: NVIDIA GeForce GTX 1080
Nvidia driver version: 465.31
cuDNN version: Probably one of the following:
/usr/local/cudnn-8.1.1-cuda-11/lib64/libcudnn.so.8.1.1
/usr/local/cudnn-8.1.1-cuda-11/lib64/libcudnn_adv_infer.so.8.1.1
/usr/local/cudnn-8.1.1-cuda-11/lib64/libcudnn_adv_train.so.8.1.1
/usr/local/cudnn-8.1.1-cuda-11/lib64/libcudnn_cnn_infer.so.8.1.1
/usr/local/cudnn-8.1.1-cuda-11/lib64/libcudnn_cnn_train.so.8.1.1
/usr/local/cudnn-8.1.1-cuda-11/lib64/libcudnn_ops_infer.so.8.1.1
/usr/local/cudnn-8.1.1-cuda-11/lib64/libcudnn_ops_train.so.8.1.1
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.5
[pip3] pytest-pytorch==0.0.0
[pip3] torch==1.9.0a0+gitda7bfa1
[conda] magma-cuda111             2.5.2                         1    pytorch
[conda] mkl                       2021.2.0           h726a3e6_389    conda-forge
[conda] mkl-include               2021.2.0           h726a3e6_389    conda-forge
[conda] numpy                     1.19.5           py36h2aa4a07_1    conda-forge
[conda] pytest-pytorch            0.2.0              pyh44b312d_0    conda-forge
[conda] torch                     1.9.0a0+gitda7bfa1           dev_0    <develop>
```

cc @mruberry @rgommers @heitorschueroff

### PR DESCRIPTION
**Pull request by pmeier**
Fixes #58988.

### CODE REVIEW

File: test/test_reductions.py [hunk 1]
```diff
@@ -2086,7 +2086,7 @@ def _compare_std_var_with_numpy(self, op, device, dtype, input, dim,
                 return
             self.fail("Failed to hit RuntimeError!")
 
-        exact_dtype = input.dtype != torch.bfloat16
+        exact_dtype = input.dtype not in (torch.bfloat16, torch.complex32, torch.complex64, torch.complex128)
```
Comment (pmeier) [inline test/test_reductions.py:2089]: Both `torch` and `numpy` convert a complex input into a float. Thus, we either ignore these cases here or we need a better mechanism to create an out tensor with the expected dtype.

**Review by pmeier — COMMENTED** [review]
All the failing examples due to a mismatching dtypes were comparisons of `torch` and `numpy` functions. Since they use different default dtypes, it should not be expected that the dtypes match if no logic is used to ensure that.

**Comment by pmeier** [comment]
There are three tests left which fail for different shapes of the input. This was not detected before, since iterating over `torch.rand((1, 1))` and `torch.rand((1, 1, 1))` or `torch.rand((1, 0))` and `torch.rand((1, 0, 42))` returns the same amount of elements.

1. `test_reductions.py::test_tensor_compare_ops_argmax_argmix_kthvalue_dim_empty`: This fails for the `kthvalue` test case:

    https://github.com/pytorch/pytorch/blob/41054f2ab5bb39d28a3eb8497f1a65b42385a996/test/test_reductions.py#L2431-L2432

    In case `np.partition` encounters an empty dimension, it does not collapse the selected axis. For example, an input array of shape `(2, 0, 4)` will stay that way. This is unlike `torch.kthvalue` which will produce `(2, 0)`. Cc @v0dro #52565

2. `quantization/core/test_quantized_op.py::test_batch_norm`: This fails for 2d inputs:

    https://github.com/pytorch/pytorch/blob/41054f2ab5bb39d28a3eb8497f1a65b42385a996/test/quantization/core/test_quantized_op.py#L2181-L2183

    It seems `torch.ops.quantized.batch_norm1d` always returns a 3d even when the input is only 2d. Cc @jerryzh168 #51597

3. `quantization/core/test_quantized_op.py::test_batch_norm_relu`: Same as above.

File: test/test_reductions.py (hunk 1 above)
Comment (mruberry) [reply test/test_reductions.py:2089]: Ignoring seems OK for now

**Review by mruberry — COMMENTED** [review]
This is cool. Nice find and great thoroughness patching the affected tests.

Should we add a test to test_testing.py to ensure this doesn't happen again? Does torch.testing.assert_close want a similar test, too?

File: test/test_reductions.py (hunk 1 above)
Comment (mruberry) [reply test/test_reductions.py:2089]: (Might as well add a TODO comment here with your thinking)

**Comment by pmeier** [comment]
> Should we add a test to test_testing.py to ensure this doesn't happen again? Does torch.testing.assert_close want a similar test, too?

`torch.testing.assert_close` has checks for mismatching dtypes

https://github.com/pytorch/pytorch/blob/6d45d7a6c331ddb856ac34a76bcd3613aa05185b/test/test_testing.py#L837-L850

as well as for `numpy` inputs

https://github.com/pytorch/pytorch/blob/6d45d7a6c331ddb856ac34a76bcd3613aa05185b/test/test_testing.py#L1088-L1094

Since we are migrating the internal test suite to use `torch.testing.assert_close` anyway (see #58981), I don't think we need to add more tests.

**Comment by pmeier** [comment]
One more shape related issue turned up: `test_lingalg.py::test_norm_matrix_degenerate_shapes` fails for 

https://github.com/pytorch/pytorch/blob/6d45d7a6c331ddb856ac34a76bcd3613aa05185b/test/test_linalg.py#L1956-L1957

with `keepdim=False` and `ord in (2, -2)`. cc @kurtamohler #42749

**Comment by mruberry** [comment]
> One more shape related issue turned up: `test_lingalg.py::test_norm_matrix_degenerate_shapes` fails for the test cases
> 
> https://github.com/pytorch/pytorch/blob/6d45d7a6c331ddb856ac34a76bcd3613aa05185b/test/test_linalg.py#L1956-L1957
> 
> with `keepdim=False` and `ord in (2, -2)`. cc @kurtamohler #42749

Sorry, would you elaborate on this failure, @pmeier, and its impact on this PR?

**Comment by pmeier** [comment]
> One more shape related issue turned up: `test_lingalg.py::test_norm_matrix_degenerate_shapes` fails for the test cases
> 
> https://github.com/pytorch/pytorch/blob/6d45d7a6c331ddb856ac34a76bcd3613aa05185b/test/test_linalg.py#L1956-L1957

Same reason as https://github.com/pytorch/pytorch/pull/59067#issuecomment-850249156: This setting should never have passed and is failing now after the comparison is working properly. 

In the cases I detailed above, the output shape of `torch.linalg.norm` deviates from `np.linalg.norm`. Since all other settings work fine, I suspect this is a bug rather than a test failure.

**Comment by mruberry** [comment]
> > One more shape related issue turned up: `test_lingalg.py::test_norm_matrix_degenerate_shapes` fails for the test cases
> > https://github.com/pytorch/pytorch/blob/6d45d7a6c331ddb856ac34a76bcd3613aa05185b/test/test_linalg.py#L1956-L1957
> 
> Same reason as [#59067 (comment)](https://github.com/pytorch/pytorch/pull/59067#issuecomment-850249156): This setting should never have passed and is failing now after the comparison is working properly.
> 
> In the cases I detailed above, the output shape of `torch.linalg.norm` deviates from `np.linalg.norm`. Since all other settings work fine, I suspect this is a bug rather than a test failure.

Got it. For each of these would you please file an issue? Can the relevant parts of each of these tests also be commented out with a link to their respective issues?

File: test/quantization/core/test_quantized_op.py [hunk 2]
```diff
@@ -2103,7 +2103,10 @@ def test_instance_norm(self):
     @skipIfNoFBGEMM
     def test_batch_norm_relu(self):
         # hypothesis too slow for this test, create test cases manually
-        max_sides = (2, 3, 4, 5)
+        # TODO: max_side==2 was disabled in https://github.com/pytorch/pytorch/pull/59067, because it failed after
+        #  fixing the underlying comparison mechanism. It should have never been passing in the first place.
+        #  This should be reinstated as soon as https://github.com/pytorch/pytorch/issues/59200 is fixed.
+        max_sides = (3, 4, 5)
```
Comment (jerryzh168) [inline test/quantization/core/test_quantized_op.py:2109]: please land after https://github.com/pytorch/pytorch/pull/59264 so that you can revert this change

**Review by mruberry — APPROVED** [review]
Nice work, @pmeier. Just drop the accidental fbgemm and kinteo changes and let's wait for CI

**Comment by mruberry** [comment]
Update: Unfortunately there are some internal tests using assertEqual that will need to be modified, too. I'll get started on them

**Comment by heitorschueroff** [comment]
@pmeier I fixed the internal failures, could you please rebase this one more time to test against new code and I'll land it after the CI runs.

**Review by heitorschueroff — APPROVED** [review]
Thanks for rebasing, LGTM.
